pynndescent/sparse_nndescent.py

```diff
@@ -70,43 +71,84 @@ def sparse_nn_descent(
-    for i in range(n_vertices):
-        indices = rejection_sample(n_neighbors, n_vertices, rng_state)
-        for j in range(indices.shape[0]):
+        c = 0
+        for i in range(n_vertices):
+            for j in range(max_candidates):
+                p = int(new_candidate_neighbors[0, i, j])
+                if p < 0:
+                    continue
+                for k in range(j, max_candidates):
+                    q = int(new_candidate_neighbors[0, i, k])
+                    if q < 0:
+                        continue
 
-            from_inds = inds[indptr[i] : indptr[i + 1]]
-            from_data = data[indptr[i] : indptr[i + 1]]
+                    from_inds = inds[indptr[p] : indptr[p + 1]]
+                    from_data = data[indptr[p] : indptr[p + 1]]
 
-            to_inds = inds[indptr[indices[j]] : indptr[indices[j] + 1]]
-            to_data = data[indptr[indices[j]] : indptr[indices[j] + 1]]
+                    to_inds = inds[indptr[q] : indptr[q + 1]]
+                    to_data = data[indptr[q] : indptr[q + 1]]
 
-            d = sparse_dist(from_inds, from_data, to_inds, to_data, *dist_args)
+                    d = sparse_dist(from_inds, from_data, to_inds, to_data, *dist_args)
 
-            heap_push(current_graph, i, d, indices[j], 1)
-            heap_push(current_graph, indices[j], d, i, 1)
-            tried.add((i, indices[j]))
-            tried.add((indices[j], i))
+                    c += heap_push(current_graph, p, d, q, 1)
+                    if p != q:
+                        c += heap_push(current_graph, q, d, p, 1)
 
-    if rp_tree_init:
-        sparse_init_rp_tree(
-            inds,
-            indptr,
-            data,
-            sparse_dist,
-            dist_args,
-            current_graph,
-            leaf_array,
-            tried=tried,
-        )
+                for k in range(max_candidates):
+                    q = int(old_candidate_neighbors[0, i, k])
+                    if q < 0:
+                        continue
 
+                    from_inds = inds[indptr[p] : indptr[p + 1]]
+                    from_data = data[indptr[p] : indptr[p + 1]]
+
+                    to_inds = inds[indptr[q] : indptr[q + 1]]
+                    to_data = data[indptr[q] : indptr[q + 1]]
+
+                    d = sparse_dist(from_inds, from_data, to_inds, to_data, *dist_args)
+
+                    c += heap_push(current_graph, p, d, q, 1)
+                    if p != q:
+                        c += heap_push(current_graph, q, d, p, 1)
+
+        if c <= delta * n_neighbors * n_vertices:
+            return
+
+
+@numba.njit(fastmath=True)
+def sparse_nn_descent_internal_high_memory(
+    current_graph,
+    inds,
+    indptr,
+    data,
+    n_vertices,
+    n_neighbors,
+    rng_state,
+    max_candidates=50,
+    sparse_dist=sparse_euclidean,
+    dist_args=(),
+    n_iters=10,
+    delta=0.001,
+    rho=0.5,
+    tried=set([(-1, -1)]),
```

Comment:
Same comment as for the non-sparse case, but just to be on the safe side, repeating myself:

Is this safe? The default `tried` will only be initialized once, so the same set will be reused if `sparse_nn_descent_internal_high_memory` is called multiple times.